It is required when using libc++. Explicitly deleting the assignment operators for `document::DocumentType::FieldSet` results in following error messages:
````

$ make documenttype.o
[ 93%] Building CXX object document/src/vespa/document/datatype/CMakeFiles/document_datatypes.dir/documenttype.cpp.o
In file included from /Users/tegge/git/vespa/document/src/vespa/document/datatype/documenttype.cpp:3:
In file included from /Users/tegge/git/vespa/document/src/vespa/document/datatype/documenttype.h:19:
/usr/local/opt/llvm/bin/../include/c++/v1/map:726:17: error: no viable overloaded '='
        __ref() = __v.__get_value();
        ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/local/opt/llvm/bin/../include/c++/v1/__tree:1662:39: note: in instantiation of member function 'std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>::operator=' requested here
            __cache.__get()->__value_ = *__first;
                                      ^
/usr/local/opt/llvm/bin/../include/c++/v1/__tree:1619:9: note: in instantiation of function template specialization 'std::__tree<std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>, std::__map_value_compare<vespalib::small_string<48>, std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>, std::less<vespalib::small_string<48>>, true>, std::allocator<std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>>>::__assign_multi<std::__tree_const_iterator<std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>, std::__tree_node<std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>, void *> *, long>>' requested here
        __assign_multi(__t.begin(), __t.end());
        ^
/usr/local/opt/llvm/bin/../include/c++/v1/map:1041:21: note: in instantiation of member function 'std::__tree<std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>, std::__map_value_compare<vespalib::small_string<48>, std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>, std::less<vespalib::small_string<48>>, true>, std::allocator<std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>>>::operator=' requested here
            __tree_ = __m.__tree_;
                    ^
/Users/tegge/git/vespa/document/src/vespa/document/datatype/documenttype.h:27:7: note: in instantiation of member function 'std::map<vespalib::small_string<48>, document::DocumentType::FieldSet>::operator=' requested here
class DocumentType : public StructuredDataType {
      ^
/usr/local/opt/llvm/bin/../include/c++/v1/__utility/pair.h:257:11: note: candidate function not viable: no known conversion from 'const std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>::value_type' (aka 'const pair<const vespalib::small_string<48>, document::DocumentType::FieldSet>') to 'const typename conditional<is_copy_assignable<first_type>::value && is_copy_assignable<second_type>::value, pair<small_string<48> &, FieldSet &>, __nat>::type' (aka 'const std::__nat') for 1st argument
    pair& operator=(typename conditional<
          ^
/usr/local/opt/llvm/bin/../include/c++/v1/__utility/pair.h:270:11: note: candidate function not viable: no known conversion from 'const std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>::value_type' (aka 'const pair<const vespalib::small_string<48>, document::DocumentType::FieldSet>') to 'typename conditional<is_move_assignable<first_type>::value && is_move_assignable<second_type>::value, pair<small_string<48> &, FieldSet &>, __nat>::type' (aka 'std::__nat') for 1st argument
    pair& operator=(typename conditional<
          ^
/usr/local/opt/llvm/bin/../include/c++/v1/__utility/pair.h:79:62: note: candidate template ignored: disabled by 'enable_if' [with _Tuple = const std::pair<const vespalib::small_string<48>, document::DocumentType::FieldSet> &]
    using _EnableB _LIBCPP_NODEBUG_TYPE = typename enable_if<_Val, bool>::type;
                                                             ^
/usr/local/opt/llvm/bin/../include/c++/v1/__utility/pair.h:42:29: note: candidate function (the implicit copy assignment operator) not viable: no known conversion from 'const pair<const std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>::key_type, std::__value_type<vespalib::small_string<48>, document::DocumentType::FieldSet>::mapped_type>' to 'const pair<vespalib::small_string<48> &, document::DocumentType::FieldSet &>' for 1st argument
struct _LIBCPP_TEMPLATE_VIS pair
                            ^
1 error generated.
make[1]: *** [document/src/vespa/document/datatype/CMakeFiles/document_datatypes.dir/documenttype.cpp.o] Error 1
make: *** [documenttype.cpp.o] Error 2
````

Note the checks for `is_copy_assignable` and `is_move_assignable`.

--- a/document/src/vespa/document/datatype/documenttype.h
+++ b/document/src/vespa/document/datatype/documenttype.h
@@ -35,7 +35,7 @@ public:
         FieldSet(const FieldSet&) = default;
         FieldSet(FieldSet&&) noexcept = default;
         FieldSet& operator=(const FieldSet&) = default;
-        FieldSet& operator=(FieldSet&&) = default;
+        FieldSet& operator=(FieldSet&&) noexcept = default;
 
         const vespalib::string & getName() const noexcept { return _name; }
         const Fields & getFields() const noexcept { return _fields; }

--- a/document/src/vespa/document/fieldset/fieldsets.h
+++ b/document/src/vespa/document/fieldset/fieldsets.h
@@ -52,10 +52,10 @@ public:
 
     FieldCollection(const DocumentType& docType, Field::Set set);
     FieldCollection(const FieldCollection &);
-    FieldCollection(FieldCollection&&) = default;
+    FieldCollection(FieldCollection&&) noexcept = default;
     ~FieldCollection() override;
     FieldCollection& operator=(const FieldCollection&) = default;
-    FieldCollection& operator=(FieldCollection&&) = default;
+    FieldCollection& operator=(FieldCollection&&) noexcept = default;
 
     bool contains(const FieldSet& fields) const override;
     Type getType() const override { return Type::SET; }

